### DISCUSSION TITLE
Sub Controller Unit Tests

### DISCUSSION THREAD

**Discussion by codeCollision4** [TheDeanLab/ASLM#377] [General]
- [x] acquire bar
- [x] autofocus popup
- [x] camera setting
- [x] gui
- [x] stage
- [ ] widget
- [ ] waveform popup
- [x] keystroke
- [x] help popup
- [x] camera view
- [ ] channels tab
- [ ] waveform tab
- [x] ilastik popup
- [ ] tiling wizard
- [ ] multi position
- [x] channel setting

**Comment by codeCollision4** [comment]
@zacsimile @annie-xd-wang @AdvancedImagingUTSW I wanted to get feedback on my plan for testing the controllers and view. I have added a dummy view and dummy controller to the conftest file. You can see it below:

    @pytest.fixture(scope='session')

    def dummy_view():

        from aslm.view.main_application_window import MainApp
    
        root = tk.Tk()
        view = MainApp(root)
        root.update()
        yield view
        root.destroy()


    @pytest.fixture(scope="package")

    def dummy_controller(dummy_model):

        from aslm.model.dummy import DummyController
    
        model = dummy_model
        controller = DummyController()
    
        controller.configuration = model.configuration
    
        return controller



I was playing around with Monkeypatching and Mocking but realized that may be overkill. My thought is that, at least for the sub controllers, we can use the dummy controller class that has the functions called by the sub controllers. 

Typically the only calls are to "execute" and the dictionary configuration attribute. The dummy controller will use the dummy model's configuration and the "execute" function will return a string. This isolates the sub-controllers from the main controller and the model.

Here is what the dummy controller looks like in code:

    class DummyController:
        def __init__(self):
            self.configuration = None
    
        def execute(str):
            return str

The dummy view will act as the Tkinter test environment. We can pass the dummy view to the sub controllers so that we can test the bindings for sub controller callbacks. This allows for the sub-controllers to be isolated from the view. 

Here is a diagram:

<img width="804" alt="image" src="https://user-images.githubusercontent.com/79557266/205713992-7c8550cb-6afd-4f7c-ac6e-263f881560e9.png">

The red circles represent what needs to be mocked, mimicked or monkey patched.

**Comment by AdvancedImagingUTSW** [comment]
Would love to hear your opinions @zacsimile and @annie-xd-wang
